### PR TITLE
soc: arm: st_stm32: add support for STM32L462xE

### PR DESCRIPTION
**Pull request by giansta**
Add support for the STM32L462xE SoC (AES not yet supported).

Signed-off-by: Giancarlo Stasi <giancarlo.stasi.co@gmail.com>

### CODE REVIEW

**Review by FRASTM — APPROVED** [review]

**Comment by giansta** [comment]
My primary e-mail wasn't public, now it should be fixed.

**Comment by ioannisg** [comment]
Is there any board, upstream that is using this SoC variant?

**Comment by giansta** [comment]
> Is there any board, upstream that is using this SoC variant?

Hi Ioannis.

Indeed I didn't find any standard boards using L462, the closest is ST "Nucleo L452RE" using L452, using STM32L452RE. The STM32L462xE has only an additional AES HW accelerator.

The board directory I'm using is part of my project, yet it's almost the same of the "Nucleo L452RE", differences are just in boards and SoC names.
Even the new files in soc/ and dts/ folders are same for L462 and L452, with just new names. AES is not even defined there, at moment they're mostly placeholders to start a project using this SoC.

Thanks,
Giancarlo.

**Comment by ioannisg** [comment]
> > Is there any board, upstream that is using this SoC variant?
> 
> Hi Ioannis.
> 
> Indeed I didn't find any standard boards using L462, the closest is ST "Nucleo L452RE" using L452, using STM32L452RE. The STM32L462xE has only an additional AES HW accelerator.
> 
> The board directory I'm using is part of my project, yet it's almost the same of the "Nucleo L452RE", differences are just in boards and SoC names.
> Even the new files in soc/ and dts/ folders are same for L462 and L452, with just new names. AES is not even defined there, at moment they're mostly placeholders to start a project using this SoC.
> 
> Thanks,
> Giancarlo.

First, thanks for contributing to Zephyr :)

So ,the problem when contributing SoC definition, but not the corresponding board definition, is that the contributed code is not tested in CI (because CI runs based on the available boards).

Now, I understand if your Board is custom; I was just thinking that if your Board is an (official) STmicro DK, you could contribute that also.

@erwango what's your thought about it?

**Comment by giansta** [comment]
> > > Is there any board, upstream that is using this SoC variant?
> > 
> > 
> > Hi Ioannis.
> > Indeed I didn't find any standard boards using L462, the closest is ST "Nucleo L452RE" using L452, using STM32L452RE. The STM32L462xE has only an additional AES HW accelerator.
> > The board directory I'm using is part of my project, yet it's almost the same of the "Nucleo L452RE", differences are just in boards and SoC names.
> > Even the new files in soc/ and dts/ folders are same for L462 and L452, with just new names. AES is not even defined there, at moment they're mostly placeholders to start a project using this SoC.
> > Thanks,
> > Giancarlo.
> 
> First, thanks for contributing to Zephyr :)
> 
> So ,the problem when contributing SoC definition, but not the corresponding board definition, is that the contributed code is not tested in CI (because CI runs based on the available boards).
> 
> Now, I understand if your Board is custom; I was just thinking that if your Board is an (official) STmicro DK, you could contribute that also.
> 
> @erwango what's your thought about it?

Yes, I confirm that it's a custom board. As far as I know, there's no DK for STM32L462xE, do you know one?

**Comment by erwango** [comment]
> @erwango what's your thought about it?

Policy is to accept SoCs not backed with an in-tree board when SoC is close enough to a SoC already present and tested in CI and the SoC doesn't enable new HW. 
There is indeed a wide range of variants, some are not present on dev boards, so we don't want to block people to push the SoCs they use on custom boards

File: dts/arm/st/l4/stm32l462Xe.dtsi [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <mem.h>
+#include <st/l4/stm32l462.dtsi>
+
+/ {
+	sram0: memory@20000000 {
+		reg = <0x20000000 DT_SIZE_K(160)>;
```
Comment (erwango) [inline dts/arm/st/l4/stm32l462Xe.dtsi]: SRAM is divided in 2 non contiguous zones (SRAM1 and SRAM2)
SRAM1 is 128Kb

**Review by erwango — CHANGES_REQUESTED** [review]

File: dts/arm/st/l4/stm32l462Xe.dtsi (hunk 1 above)
Comment (giansta) [reply dts/arm/st/l4/stm32l462Xe.dtsi]: The datasheet (https://www.st.com/resource/en/datasheet/stm32l462ve.pdf), section 3.5, states:
" 32 Kbyte located at address 0x1000 0000 with hardware parity check (SRAM2).
This memory is also mapped at address 0x2002 0000, offering a contiguous address 
space with the SRAM1 (32 Kbyte aliased by bit band)"

It's same for L452 and L462.

Comment (erwango) [reply dts/arm/st/l4/stm32l462Xe.dtsi]: @giansta my bad I have missed this mapping. 160Kb it is then.

**Review by erwango — APPROVED** [review]

**Review by mnkp — APPROVED** [review]
Verified on a custom board.

**Review by ioannisg — APPROVED** [review]

**Pull request merged**